### ISSUE TITLE
Master vs 19.10 for Xavier deployment 

### ISSUE THREAD

**Issue by mbufi** [NVIDIA/retinanet-examples#198]
Hello,
Thanks for the great repo. Documentation is fantastic. I am unsure of one thing though...

My end goal is to deploy my models to Xavier. I know to train models I read that I need to use the 19.10 branch but the master branch has more optional training arguments in the documentation. If I'm going to be training on a host computer can I not use the master branch and then export to Onnx for the Xavier ? 

Could you please let me know what are the differences between the two branches besides tensorrt version? Is all the training code the same ? Are there any other differences?  I'm worried of missing features from the master branch with using 19.10. 
Also, do you have official Xavier benchmarks for inference speed? 

Thank you in advance!

**Comment by yashnv** [comment]
Hello, yes you can use master for training and exporting to ONNX, then switch to 19.10 branch on Xavier for TRT export.

The major difference is:

- Plugin update to IPluginV2Ext from IPluginV2 to support explicit batching which becomes a requirement in TRT7. Most of the other changes revolve around this.

Training code is mostly the same for Axis-aligned bounding box training, apart from some data augmentation options.
There isn't an official benchmark yet. For max performance on Xavier, you need to enable the MAXP mode.

**Comment by mbufi** [comment]
@yashnv thank you for your quick response! Shows your dedication to the repo. 

For the tensorrt network generation, I would bring my exported Onnx model to my Xavier --> clone the repo --> checkout 19.10 and install everything --> convert onmxodel to tensorrt engine. Correct?

**Comment by yashnv** [comment]
No, you don't need to pip install. You just need to go to extras->cppapi->build export and infer.

**Comment by mbufi** [comment]
Fantastic. Thakns for the help @yashnv !

**Comment by mbufi** [comment]
Status update:
Training the model with the master branch and then trying to use 19.10 to convert to tensorRT does not work.

> nvidia@nvidia:~/ai/retinanet-examples/extras/cppapi/build$ ./export model.onnx engine.plan
> Building engine...
> Building FP16 core model...
> Building accelerated plugins...
> Parameter check failed at: ../builder/Network.cpp::addConcatenation::197, condition: nbInputs > 0 && nbInputs < MAX_CONCAT_INPUTS
> Segmentation fault (core dumped)

What should I do? 

Thanks!

**Comment by yashnv** [comment]
What was your exact training command on master?
Where did you export the ONNX model? master or 19.10?

**Comment by mbufi** [comment]
commands in order:
1. 
```sudo docker run --gpus all --rm --ipc=host -it -v/home/musashi/ai/dataset_converter/test_data/coco:/tmp/trainingData odtk:latest```

2.  Change the scales in the model.py to train for smaller objects. Works good!
model.py
` self.scales = [1, 2, 4]`
`pip install -e .`

3. Fine-tune network 
```
odtk train /tmp/trainingData/model_mydataset.pth --backbone ResNet50FPN \
    --fine-tune /tmp/trainingData/retinanet_rn50fpn.pth \
    --classes 4 --iters 10000 --val-iters 1000 --lr 0.0005 --augment-rotate --batch 4 \
    --resize 1024 --jitter 720 960 --images /tmp/trainingData/images/ \
    --annotations /tmp/trainingData/annotations/train.json --val-annotations /tmp/trainingData/annotations/val.json
```

4. Export the pth model to onnx for xavier
`odtk export /tmp/trainingData/model_mydataset.pth /tmp/trainingData/model.onnx --size 800 1280`

5. Transfer `model.onnx` to xavier and build the cpp api to export 
6. export 
`./export model.onnx engine.plan`

**Comment by mbufi** [comment]
Also: Xavier is jetpack ver:  4.2.3

**Comment by mbufi** [comment]
**Update**: I just built my own docker container for 19.10 and trained a model. I was then able to successfully transfer its onnx model to the Xavier to create a tensorRT engine to perform inference. 

Therefore i have confirmed that  **training on Master branch** --> **inference on Xavier with 19.10** 
does not work. 

Only **training on 19.10 branch** --> **inference on Xavier with 19.10** works. 

Thanks!

**Comment by batrlatom** [comment]
I came across the problem literary hour after your last post. I also tried master -> 19.10 way without success, but 19.10 -> 19.10 does not work for me too. I Think that you also need to have jetpack 4.3 and not 4.4 installed

**Comment by mbufi** [comment]
@batrlatom I am using jetpack 4.2.3 on purpose because i know minor things change in 4.3 and 4.4

I can confirm 19.10 --> Xavier with 4.2.3 works.

**Comment by batrlatom** [comment]
@mbufi Thanks. I was able to run it with Jetpack 4.3 . Just one more question. What is your inference time with standard resnet50 on ( 15 - 30W )? I am getting something like 120-140 ms.

**Comment by yashnv** [comment]
Looks like the ONNX Parser in Jetson needs to be updated to correctly parse the ONNX model trained through master branch.

@mbufi there's a workaround, but it is complicated:

- Setup a 19.10 docker on host machine, then use your trained .pth file to export to ONNX model on the 19.10 docker on your host. **IMPORTANT: Change the scales, ratios to the same setting as you used in the master 20.02 container in the model.py file**

- Then finally use that exported ONNX model in your Jetson to export to plan. Hopefully, this should work.

**Comment by mbufi** [comment]
@yashnv  Thank you for your suggestions. 

Can confirm that this works.

**Comment by anurag8786** [comment]
Is there any way, to convert onnx file to plan file in NVidia Jetson Xavier NX, since the minimum downloadable version of Jetpack is 4.4 for Xavier NX
